envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataValidatorTest.kt

```diff
@@ -69,6 +92,63 @@ class NodeMetadataValidatorTest {
         // no exception thrown
     }
 
+    @ParameterizedTest
+    @MethodSource("configurationModeNotSupported")
+    fun `should fail if service want to use mode which server doesn't support`(
+        adsSupported: Boolean,
+        xdsSupported: Boolean,
+        ads: Boolean,
+        modeNotSupportedName: String
+    ) {
+        // given
+        val configurationModeValidator = NodeMetadataValidator(SnapshotProperties().apply {
+            configurationMode = createConfigurationMode(ads = adsSupported, xds = xdsSupported)
+        })
+
+        val node = node(
+            ads = ads,
+            serviceDependencies = setOf("a", "b", "c"),
+            serviceName = "regular-1"
+        )
+        val request = DiscoveryRequest.newBuilder().setNode(node).build()
+
+        // expects
+        assertThatExceptionOfType(ConfigurationModeNotSupportedException::class.java)
+            .isThrownBy { configurationModeValidator.onStreamRequest(streamId = 123, request = request) }
+            .satisfies {
+                assertThat(it.status.description).isEqualTo(
+                    "Blocked service regular-1 from receiving updates. $modeNotSupportedName is not supported by server."
+                )
+                assertThat(it.status.code).isEqualTo(Status.Code.INVALID_ARGUMENT)
+            }
+    }
+
+    @ParameterizedTest
+    @MethodSource("configurationModeSupported")
+    fun `should do nothing if service want to use mode which server supports`(
+        adsSupported: Boolean,
+        xdsSupported: Boolean,
+        ads: Boolean
+    ) {
+        // given
+        val configurationModeValidator = NodeMetadataValidator(SnapshotProperties().apply {
+            configurationMode = createConfigurationMode(ads = adsSupported, xds = xdsSupported)
+        })
+
+        val node = node(
+            ads = ads,
+            serviceDependencies = setOf("a", "b", "c"),
+            serviceName = "regular-1"
+        )
+        val request = DiscoveryRequest.newBuilder().setNode(node).build()
+
+        // when
+        configurationModeValidator.onStreamRequest(123, request = request)
+
+        // then
+        // no exception thrown
```

Comment:
assertDoesNotThrow? (you'll get a nice error message instead of crashing the test suite)